Match/exclude destination rules (flattened) (#281)
tfpauly/privacy-proxy: draft-ietf-intarea-proxy-config.md

only keys that are registered in an IANA registry, defined in {{proxy-info-iana}} and that are marked
as optional.  The `mandatory` list MUST NOT include any entries that are not present in the sub-dictionary.

> The value of `identifier` key is an optional string that can be used to refer to the

Reviewer: ```suggestion
The value of `identifier` key is an optional string that can be used to refer to a particular
```

proxy from other dictionaries, specifically those defined in {{destinations}}. The
string value is an arbitrary JSON string. Identifier values MAY be duplicated

> across different proxy dictionaries in the `proxies` array, which would indicate

Reviewer: ```suggestion
across different proxy dictionaries in the `proxies` array, which indicates
```

> that all references from other dictionaries to a particular identifier value applies

Reviewer: ```suggestion
that all references from other dictionaries to a particular identifier value apply
```

a proxy. Destination restrictions are often used in VPN tunnel configurations such as split
DNS in IKEv2 {{?IKEV2SPLIT=RFC8598}}.

PvD Additional Information can be used to indicate that a set of proxies only allows access to
a limited set of destinations.

To support determining which traffic is supported by different proxies, this document defines

> a new PvD Additional Information key `proxy-match`. This key has value that is arrays of

Reviewer: ```suggestion
a new PvD Additional Information key `proxy-match`. This key has a value that is an array of
```

dictionaries, where each subdictionary describes a rule for matching traffic to one or more

> proxies or excluding the traffic from all proxies. These subdictionaries are referred

Reviewer: ```suggestion
proxies, or excluding the traffic from all proxies described in the PvD. These subdictionaries are referred
```

to as "destination rules", since they define rules about which destinations can be accessed
for a particular proxy.

> ## Destination rule keys

Reviewer: ```suggestion
## Destination Rule Keys
```

This document defines four keys for destination rules. Any destination rule MUST contain

> the `proxies` key. Value corresponding to the `proxies` key may be either an empty array

Reviewer: ```suggestion
the `proxies` key. Values corresponding to the `proxies` key may be either an empty array,
```

> which implies that no proxy defined in this PvD can process matching traffic or an array

Reviewer: ```suggestion
which implies that no proxy defined in this PvD can process matching traffic, or an array of strings
```

> with at least one proxy `identifier`. All destination rules MUST also contain at least one

Reviewer: ```suggestion
with at least one proxy `identifier` string. All destination rules MUST also contain at least one
```

in the `domains` array of a `proxy-match` rule.

The `subnets` array includes IPv4 and IPv6 address literals, as well as IPv4 and IPv6 address subnets
written using CIDR notation. Subnet-based destination information are only meant be used when applications
are communicating with destinations identified by only an IP address, and not a hostname.

Destination rules SHOULD NOT contain both the `domains` key and the `subnets` key. Many
clients will not resolve a domain being accessed through the proxy to an IP address before using
a proxy, so the subnet information may not be available.

The `ports` array includes specific ports (used for matching TCP and/or UDP ports), as well as
ranges of ports written with a low port value and a high port value, with a `-` in between.
For example, "1024-2048" matches all ports from 1024 to 2048, including the 1024 and 1028.
If `ports` key is not present, all ports are assumed to match. Comma-separated port list may
contain individual port numbers (such as "80") or inclusive ranges of ports. For example
"1024-2048" matches all ports from 1024 to 2048, including the 1024 and 1028.

> ##  Using destination rules

Reviewer: ```suggestion
##  Using Destination Rules
```

and the client has validated that that the hostname of the proxy, the identifier of
the PvD, and the validated hostname identity on the certificate all match.

When using information in destination rules ({{destinations}}), clients MUST only allow
the PvD configuration to narrow the scope of traffic that they will send through a proxy.
Clients that are configured by policy to only send a particular set of traffic through
a particular proxy can learn about rules that will cause them to send more narrowly-scoped
traffic, but MUST NOT send traffic that would go beyond what is allowed by local policy.

# IANA Considerations

## New PvD Additional Information key {#proxies-key-iana}

This document registers two new keys in the "Additional Information PvD Keys" registry.

> ### proxies Key

Reviewer: ```suggestion
### `proxies` Key
```

  "proxy": "https://proxy.example.org/masque{?target_host,target_port}"
} ]

> ### proxy-match Key

Reviewer: ```suggestion
### `proxy-match` Key
```
